azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -59,95 +64,152 @@ def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
-        role_instance = find(xml_doc, "RoleInstance")
+        # Retrieve goal state information
+        self.incarnation = findtext(self._xml_doc, "Incarnation")
+        role_instance = find(self._xml_doc, "RoleInstance")
         self.role_instance_id = findtext(role_instance, "InstanceId")
         role_config = find(role_instance, "Configuration")
         self.role_config_name = findtext(role_config, "ConfigName")
-        container = find(xml_doc, "Container")
+        container = find(self._xml_doc, "Container")
         self.container_id = findtext(container, "ContainerId")
-        lbprobe_ports = find(xml_doc, "LBProbePorts")
+        lbprobe_ports = find(self._xml_doc, "LBProbePorts")
         self.load_balancer_probe_port = findtext(lbprobe_ports, "Port")
-
         GoalState.ContainerID = self.container_id
 
-        fetch_full_goal_state = False
-        if full_goal_state:
-            fetch_full_goal_state = True
-            reason = 'force update'
-        elif base_incarnation is not None and self.incarnation != base_incarnation:
-            fetch_full_goal_state = True
-            reason = 'new incarnation'
-
-        if not fetch_full_goal_state:
-            self.hosting_env = None
-            self.shared_conf = None
-            self.certs = None
-            self.ext_conf = None
-            self.remote_access = None
-            return
+        if retrieval_mode != GoalStateRetrievalMode.Limited:
+            # Read properties necessary to retrieve the FastTrack goal state
+            self._retrieve_fabric_ext_conf_properties()
 
-        logger.info('Fetching new goal state [incarnation {0} ({1})]', self.incarnation, reason)
+            # Retrieve the extension config, which we need to determine if anything changed
+            self._retrieve_ext_conf()
 
-        try:
-            uri = findtext(xml_doc, "HostingEnvironmentConfig")
-            xml_text = wire_client.fetch_config(uri, wire_client.get_header())
-            self.hosting_env = HostingEnv(xml_text)
+            if self.ext_conf is not None or retrieval_mode == GoalStateRetrievalMode.Full:
+                # We only need to retrieve the certificates if there's a Fabric incarnation change.
+                # FastTrack doesn't affect them
+                if self.ext_conf.is_fabric_change or retrieval_mode == GoalStateRetrievalMode.Full:
+                    self._retrieve_certificates()
+
+                # Retrieve the other documents if we have either a Fabric or FastTrack change
+                if self.ext_conf.changed or retrieval_mode == GoalStateRetrievalMode.Full:
+                    self._retrieve_hosting_env()
+                    self._retrieve_shared_conf()
+                    self._retrieve_remote_access()
+
+    @staticmethod
+    def fetch_goal_state(wire_client, ext_config_retriever):
+        """
+        Fetches the goal state, not including any nested properties (such as remote access).
```

Comment:
this one, i think, will fetch the full goal state if it changed

Reply:
Fixed the comment